pkg/storage/iterators.go

```diff
@@ -253,6 +339,46 @@ func (f *ConditionsFilteredTupleKeyIterator) Stop() {
 	f.iter.Stop()
 }
 
+// Head returns the next most tuple in the underlying iterator that meets
+// the filter function this iterator was constructed with.
+// The underlying iterator will not advance.
+// This function is not thread-safe.
+func (f *ConditionsFilteredTupleKeyIterator) Head(ctx context.Context) (*openfgav1.TupleKey, error) {
+	for {
+		tuple, err := f.iter.Head(ctx)
+		if err != nil {
+			if errors.Is(err, ErrIteratorDone) {
+				if f.onceValid || f.lastError == nil {
+					return nil, ErrIteratorDone
+				}
+				return nil, f.lastError
+			}
+			return nil, err
+		}
+
+		valid, err := f.filter(tuple)
+		if err != nil {
+			f.lastError = err
+			_, err = f.iter.Next(ctx)
+			if err != nil {
+				// should never happen except if the underlying ds has error
+				return nil, err
+			}
+			continue
+		}
+		if !valid {
+			_, err = f.iter.Next(ctx)
+			if err != nil {
+				// should never happen except if the underlying ds has error
```

Comment:
Ditto, since you called `Next`

Reply:
See above, the head should have produced the Done already.